src/sckmerdb_build.cpp

```diff
@@ -5,403 +5,441 @@
-    int_type seq_code = 0;
+template <int len>
+uint64_t seq_encode(const char* buf) {
+    constexpr int BITS_PER_BASE = 2;
+    uint64_t seq_code = 0;
     for (int i=0;  i < len;  ++i) {
-        const int_type b_code = code_dict[buf[i]];
-        seq_code |= ((b_code & b_mask) << (bpb * (len - i - 1)));
+        const auto b_code = code_dict.data[buf[i]];
+        assert(b_code != -1 && "Trying to encode a character that is not ACTG or actg.");
+        seq_code |= (((uint64_t) b_code) << (BITS_PER_BASE * i));
     }
     return seq_code;
 }
 
-template <class int_type>
-void seq_decode(char* buf, const int len, const int_type seq_code, int_type* code_dict, const int_type b_mask) {
-    for (int i=0;  i < len-1;  ++i) {
-        const int_type b_code = (seq_code >> (bpb * (len - i - 2))) & b_mask;
-        buf[i] = bit_decode<int_type>(b_code);
-    }
-
-    buf[len-1] = '\0';
-}
-
 
-template <class int_type>
-void bit_load(const char* k_path, vector<char>& buffer, vector<tuple<int_type, int_type>>& k_vec, const int_type* code_dict, const int_type b_mask) {
-    auto t_start = chrono_time();
-
-    char* window = buffer.data();
-
-    uintmax_t n_lines = 0;
+// INPUT:  A sckmers.tsv file with content like:
+//
+// ...
+// 343 25  TGGATACCACGGCGCAAGAGCACGCACAGAA TGGATACCACGGCGCAAGAGCACGCGCAGAA TTCTGTGCGTGCTCTTGCGCCGTGGTATCCA TTCTGCGCGTGCTCTTGCGCCGTGGTATCCA 1   11  259564  8   3
+// 343 24  GGATACCACGGCGCAAGAGCACGCACAGAAT GGATACCACGGCGCAAGAGCACGCGCAGAAT ATTCTGTGCGTGCTCTTGCGCCGTGGTATCC ATTCTGCGCGTGCTCTTGCGCCGTGGTATCC 1   11  259564  8   3
+// ...
+//
+// where each line contains, in this order:
+//
+//      $1 - genomic position of SNP
+//      $2 - zero-based offset of SNP in forward 31-mer
+//      $3 - forward 31-mer covering the SNP's major allele
+//      $4 - forward 31-mer covering the SNP's minor allele
+//      $5 - reverse complement of $3
+//      $6 - reverse complement of $4
+//      $7 - unused
+//      $8 - unused
+//      $9 - 6 decimal digit species ID
+//      $10 - unused
+//
+//  a "snp coordinate" is obtained by concatenating the following character strings and
+//  converting the result back to an integer:
+//
+//      * the 6 decimal digit species id (input column $8),
+//
+//      * a single digit allele type:
+//           (0 = major allele, matching input $3 and $5,
+//            1 = minor allele, matching input $4 and $6),
+//
+//      * genomic position of SNP (input column $1)
+//
 
-    int fd;
-    fd = open(k_path, O_RDONLY);
 
-    int cur_pos = 0;
-    int snp_pos = 0;
 
-    char seq_buf[k];
-	char snp_id[16];
+void bit_load(const char* k_path, vector<KmerData>& result) {
 
-    //auto fh = fstream(out_path, ios::out | ios::binary);
+    assert(k_path);
+    FILE *fp = fopen(k_path, "r");
+    if (fp == NULL) {
+        cerr << "Trouble opening file " << k_path << "." << endl;
+        exit(EXIT_FAILURE);
+    }
 
-	bool id_switch = false;
-    bool has_wildcard = false;
+    // 3 columns in each line.
```

Comment:
I now realize this comment is out of date and should just be deleted.   We actually have 10 columns as explained in the more detailed comment above.